### ISSUE TITLE
Error with Prisma Accelerate at the edge using unsupported Nodejs api

### ISSUE THREAD

**Issue by nnennajohn** [prisma/prisma#21463]
### Bug description

When I try to run `@prisma/client/edge` with `@prisma/extension-accelerate` in nextjs middleware, I get an error about it using a Node API not yet supported in the Edge Runtime. See screenshot.

But the log after running `prisma generate --no-engine` suggests that you can use accelerate at the edge and points you to the accelerate docs.

Can it not yet be used at the edge and is the console printout an error?

Thank you.

#### Next JS Error
<img width="699" alt="Screenshot 2023-10-11 at 11 52 19 PM" src="https://github.com/prisma/prisma/assets/1761197/c7d04094-9157-425c-abf4-9f4f44d80f15">

#### Edge Prisma declaration
<img width="763" alt="Screenshot 2023-10-11 at 11 51 33 PM" src="https://github.com/prisma/prisma/assets/1761197/56dfb7eb-373f-4071-80d8-253a7d3b90f0">

#### Usage in middleware
<img width="662" alt="Screenshot 2023-10-12 at 12 05 41 AM" src="https://github.com/prisma/prisma/assets/1761197/25c1c80f-b5b0-4cc8-bf38-428d25cf686b">

#### Console print suggesting you can use accelerate in edge environments
<img width="658" alt="Screenshot 2023-10-12 at 12 07 11 AM" src="https://github.com/prisma/prisma/assets/1761197/83147a1b-92b0-47e1-bacf-913be6d267a7">

### How to reproduce

Create new prisma client with edge and accelerate.

Use in it a nextjs middleware build

Build the nextjs project.

### Expected behavior

No edge runtime error about using an unsupported Node.js API

<img width="772" alt="Screenshot 2023-10-12 at 12 09 42 AM" src="https://github.com/prisma/prisma/assets/1761197/c78affaa-020e-4975-92b6-66bcfd3905da">

### Prisma information

```prisma
// Add your schema.prisma

datasource db {
  provider  = "postgresql"
  url       = env("DATABASE_URL")
  directUrl = env("DATABASE_DIRECT_URL")
}

generator client {
  provider = "prisma-client-js"
  previewFeatures = ["fullTextSearch"]
}

model UserProfile {
  id                              String                     @id @default(uuid()) @db.Uuid
  createdAt                       DateTime?                  @default(now())
  updatedAt                       DateTime?                  @default(now()) @updatedAt
  deletedAt                       DateTime?
  email                           String                     @unique
  authUserId                      String?                    @unique
}
```

```ts
// Add your code using Prisma Client

import { PrismaClient } from "@prisma/client/edge";
import { withAccelerate } from "@prisma/extension-accelerate";

const prismaClientEdge = new PrismaClient().$extends(withAccelerate());
```

### Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: v18.17.1

### Prisma Version

```
prisma                  : 5.4.2
@prisma/client          : 5.4.2
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine ac9d7041ed77bcc8a8dbd2ab6616b39013829574 (at ../../node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli ac9d7041ed77bcc8a8dbd2ab6616b39013829574 (at ../../node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.4.1-2.ac9d7041ed77bcc8a8dbd2ab6616b39013829574
Default Engines Hash    : ac9d7041ed77bcc8a8dbd2ab6616b39013829574
Studio                  : 0.494.0
Preview Features        : fullTextSearch
```

**Comment by janpio** [comment]
Yes, generally that should work - but maybe we missed some case here. Thanks for reporting this!

Can you maybe create a minimal Next.js project with the setup you describe which shows this problem, and then put it on GitHub? That would make it the easiest for us to reproduce, otherwise there are quite a few steps where we might go wrong and mess it up. Thanks!

**Comment by janpio** [comment]
We had a first look, and the problem actually seems to be caused by the `@prisma/extension-accelerate` codebase - you can also see that in the error message above. I notified the correct team and they will look into that soon.

**Comment by nnennajohn** [comment]
Great. I was just pondering how I'd go about making a repro, given that the accelerate service is hosted, and I'd have to put up a hosted db, shared that with accelerate to get an api key, etc.

Glad you tracked down the problem. I appreciate you looking into it.

Do I keep this open in the mean time?

**Comment by JoelVenable** [comment]
Also running into this issue. Any updates?

**Comment by nurul3101** [comment]
@JoelVenable Our team is aware of this issue. I'll post a message here as soon as there is an update. 🙏

**Comment by Jolg42** [comment]
Internal notes
- https://github.com/prisma/prisma-extension-accelerate/blob/993200927b163728cc2380972386e01884cbc61b/src/user-agent.ts#L27
This is likely the code involved here 
```
/**
 * Generates a User-Agent segment for the JavaScript runtime environment.
 */
function getRuntimeSegment(): string {
  if (typeof navigator !== "undefined") {
    // Deno, Bun, Cloudflare Workers, general WinterCG compat
    return navigator.userAgent;
  } else if (
    typeof process !== "undefined" &&
    typeof process.versions !== "undefined"
  ) {
    return `Node/${process.versions.node} (${process.platform}; ${process.arch})`;
  } else if ("EdgeRuntime" in globalThis) {
    return `Vercel-Edge-Runtime`;
  } else {
    return `UnknownRuntime`;
  }
}
```
- https://prisma-company.slack.com/archives/C042HLMAC5U/p1705941372897809
> We try to access that Node API conditionally but Next.js does weird static analysis checks like this that are unavoidable. We'll probably need yet another hacky solution to hide this from Next.js's build system.

**Comment by cyrus-za** [comment]
@Jolg42 any update on this? I am also running into it. It's a little weird that the accelerate extension package would use process.versions as its a node api that does not run on the edge and the whole point of accelerate is to run on the edge, right?

**Comment by Jolg42** [comment]
So the code actually only uses `process.versions` if it is defined.
The problem is static analysis from Next.js here, which is technically incorrect, since it will not be used at runtime.
The team in charge already knows about this, I'm unsure what they want to do for this.

To me, I think it would be a good idea to report it to Next.js to tweak the static analysis.
Source code for the error: https://github.com/vercel/next.js/blob/e273b335da63b13331ffd367914fb9fbd8adc06c/packages/next/src/build/webpack/plugins/middleware-plugin.ts#L345-L366

**Comment by cyrus-za** [comment]
Ok I reported it [here](https://github.com/vercel/next.js/issues/64143)

Let's see what they say

**Comment by ruanosena** [comment]
Same bug on build, tried to use it with next-auth (auth.js). In production had this error `Error running the exported Web Handler: [Error: No fetch event listeners found]` in the edge middleware calls. I opted-out of accelerate and error was gone.

**Comment by igalklebanov** [comment]
Hey 👋 

Closing as its rusty and dusty.
